torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,390 @@
+    case ::at::ScalarType::Float: func<float>(args); break;             \
+    case ::at::ScalarType::Double: func<double>(args); break;           \
+    case ::at::ScalarType::Half: func<gloo::float16>(args); break;      \
+    case ::at::ScalarType::Char: func<int8_t>(args); break;             \
+    case ::at::ScalarType::Byte: func<uint8_t>(args); break;            \
+    case ::at::ScalarType::Int: func<int32_t>(args); break;             \
+    case ::at::ScalarType::Long: func<int64_t>(args); break;            \
+    default:                                                            \
+      throw std::runtime_error("Invalid scalar type");                  \
+  }
+
+namespace c10d {
+
+using KeyType = AlgorithmKey;
+using EntryType = std::unique_ptr<AlgorithmEntry>;
+
+namespace {
+
+// Wrap c10d store as Gloo store
+class GlooStore : public ::gloo::rendezvous::Store {
+ public:
+  GlooStore(const std::shared_ptr<::c10d::Store>& store)
+      : store_(store) {
+  }
+
+  void set(const std::string& key, const std::vector<char>& value) override {
+    // @pietern was unable to figure out a way to cast the value of
+    // type std::vector<char> to an std::vector<unsigned char>,
+    // so going with a copy here (it's not a hot path so it's fine).
+    std::vector<uint8_t> tmp(value.size());
+    memcpy(tmp.data(), value.data(), value.size());
+    store_->set(key, tmp);
+  }
+
+  std::vector<char> get(const std::string& key) override {
+    // @pietern was unable to figure out a way to cast the value of
+    // type std::vector<unsigned char> to an std::vector<char>,
+    // so going with a copy here (it's not a hot path so it's fine).
+    auto value = store_->get(key);
+    std::vector<char> tmp(value.size());
+    memcpy(tmp.data(), value.data(), value.size());
+    return tmp;
+  }
+
+  void wait(const std::vector<std::string>& keys) override {
+    store_->wait(keys, Store::kDefaultTimeout);
+  }
+
+  void wait(
+      const std::vector<std::string>& keys,
+      const std::chrono::milliseconds& timeout) override {
+    store_->wait(keys, timeout);
+  }
+
+ protected:
+  std::shared_ptr<::c10d::Store> store_;
+};
+
+template <typename T>
+const ::gloo::ReductionFunction<T>* reductionFunction(const ReduceOp& r) {
+  switch (r) {
+    case ReduceOp::SUM:
+      return ::gloo::ReductionFunction<T>::sum;
+    case ReduceOp::PRODUCT:
+      return ::gloo::ReductionFunction<T>::product;
+    case ReduceOp::MIN:
+      return ::gloo::ReductionFunction<T>::min;
+    case ReduceOp::MAX:
+      return ::gloo::ReductionFunction<T>::max;
+  }
+
+  throw std::runtime_error("Unhandled ReduceOp");
+}
+
+} // namespace
+
+ProcessGroupGloo::WorkGloo::WorkGloo() {
+  completed_ = false;
+}
+
+ProcessGroupGloo::WorkGloo::~WorkGloo() {
+}
+
+bool ProcessGroupGloo::WorkGloo::isCompleted() {
+  std::unique_lock<std::mutex> lock(m_);
+  return completed_;
+}
+
+bool ProcessGroupGloo::WorkGloo::wait() {
```

Comment:
const?

Reply:
Does a mutex grab + condvar wait, so doubt it can be